### PR TITLE
services/prometheus/exporters: add scaphandre

### PR DESCRIPTION
**Pull request by gaelreyrol**
###### Description of changes

This PR add scaphandre package to prometheus exporters.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/monitoring/prometheus/exporters.nix [hunk 1]
```diff
@@ -301,6 +302,13 @@ in
         Please specify either 'services.prometheus.exporters.sql.configuration' or
           'services.prometheus.exporters.sql.configFile'
       '';
+    } {
+      assertion = cfg.scaphandre.enable -> (
+        (builtins.elem "intel_rapl_common" config.boot.kernelModules)
+      );
+      message = ''
+        Please enable 'intel_rapl_common' in boot.kernelModules
+      '';
```
Comment (gaelreyrol) [inline nixos/modules/services/monitoring/prometheus/exporters.nix:311]: I think I should also assert that machine runs on Intel or AMD according to the [compatibility](https://hubblo-org.github.io/scaphandre-documentation/compatibility.html) page but I don't know how to do that.

File: nixos/tests/prometheus-exporters.nix [hunk 2]
```diff
@@ -1085,6 +1085,22 @@ let
       '';
     };
 
+    scaphandre = {
+      exporterConfig = {
+        enable = true;
+      };
+      metricProvider = {
+        boot.kernelModules = [ "intel_rapl_common" ];
```
[the remark this answers is not in the document]
Comment (gaelreyrol) [reply nixos/tests/prometheus-exporters.nix:1093]: Well it produces an error but is not what's tested, should I try to handle it?
image

File: nixos/modules/services/monitoring/prometheus/exporters.nix (hunk 1 above)
Comment (gaelreyrol) [reply nixos/modules/services/monitoring/prometheus/exporters.nix:319]: I tried to restrict as much as possible the possibility to use Scaphandre without having the proper configuration.

File: nixos/modules/services/monitoring/prometheus/exporters/scaphandre.nix [hunk 3]
```diff
@@ -0,0 +1,31 @@
+{ config, lib, pkgs, options }:
+
+with lib;
```
Comment (drupol) [inline nixos/modules/services/monitoring/prometheus/exporters/scaphandre.nix:3]: The trends in `nixpkgs` is to remove this from modules.

**Comment by gaelreyrol** [comment]
@GrahamcOfBorg build

**Review by drupol — CHANGES_REQUESTED** [review]

File: nixos/modules/services/monitoring/prometheus/exporters/scaphandre.nix (hunk 3 above)
Comment (drupol) [reply nixos/modules/services/monitoring/prometheus/exporters/scaphandre.nix]: Thanks for fixing it so promptly :)

**Review by drupol — APPROVED** [review]
LGTM

File: nixos/modules/services/monitoring/prometheus/exporters.nix [hunk 4]
```diff
@@ -301,6 +302,21 @@ in
         Please specify either 'services.prometheus.exporters.sql.configuration' or
           'services.prometheus.exporters.sql.configFile'
       '';
+    } {
+      assertion = cfg.scaphandre.enable -> (pkgs.stdenv.hostPlatform.isx86_64 == true);
+      message = ''
+        Only x86_64 host platform architecture is not supported.
```
Comment (ambroisie) [inline nixos/modules/services/monitoring/prometheus/exporters.nix:308]: I think you meant

```suggestion
        Only x86_64 host platform architecture is supported.
```

Comment (ambroisie) [reply nixos/modules/services/monitoring/prometheus/exporters.nix:308]: And are you sure the (build) host should be x86_64? Or do you mean the (build) target, i.e: the machine that will run the code?

File: nixos/modules/services/monitoring/prometheus/exporters.nix [hunk 5]
```diff
@@ -301,6 +302,21 @@ in
...
+      '';
+    } {
+      assertion = cfg.scaphandre.enable -> ((lib.kernel.whenHelpers pkgs.linux.version).whenOlder "5.11" true).condition == false;
+      message = ''
+        A kernel version newer than '5.11' is required. ${pkgs.linux.version}
```
Comment (ambroisie) [inline nixos/modules/services/monitoring/prometheus/exporters.nix:313]: This should explain that the assertion comes from `cfg.scaphandre.enable`.

File: nixos/modules/services/monitoring/prometheus/exporters.nix [hunk 6]
```diff
@@ -301,6 +302,21 @@ in
...
+      '';
+    } {
+      assertion = cfg.scaphandre.enable -> (builtins.elem "intel_rapl_common" config.boot.kernelModules);
+      message = ''
+        Please enable 'intel_rapl_common' in 'boot.kernelModules'.
```
Comment (ambroisie) [inline nixos/modules/services/monitoring/prometheus/exporters.nix:318]: Ditto.

**Review by ambroisie — COMMENTED** [review]
I got linked [to the commit](https://github.com/NixOS/nixpkgs/commit/e1b3f7b15902d1f1b7560fd610f98a24821e1759) from Matrix, I think there are some issues with this MR as it was merged 🙃.

**Comment by drupol** [comment]
> I got linked [to the commit](https://github.com/NixOS/nixpkgs/commit/e1b3f7b15902d1f1b7560fd610f98a24821e1759) from Matrix, I think there are some issues with this MR as it was merged upside_down_face.

Still learning the job of a maintainer every single day! Sorry for the trouble !

File: nixos/modules/services/monitoring/prometheus/exporters.nix (hunk 4 above)
Comment (gaelreyrol) [reply nixos/modules/services/monitoring/prometheus/exporters.nix:308]: Well I mean the build target indeed.

File: nixos/modules/services/monitoring/prometheus/exporters.nix (hunk 5 above)
Comment (gaelreyrol) [reply nixos/modules/services/monitoring/prometheus/exporters.nix:313]: Will do

**Comment by gaelreyrol** [comment]
@ambroisie What are the issues other than your review comments?

**Comment by ambroisie** [comment]
@gaelreyrol I meant the ones I commented on :-).

**Pull request merged**